schema/pgsql/schema.sql

```diff
@@ -124,21 +124,34 @@ CREATE TABLE schedule_member (
 
 CREATE TABLE source (
     id bigserial,
+    -- The type "icinga2" is special and requires (at least some of) the icinga2_ prefixed columns.
     type text NOT NULL,
     name text NOT NULL,
     -- will likely need a distinguishing value for multiple sources of the same type in the future, like for example
     -- the Icinga DB environment ID for Icinga 2 sources
 
-    -- listener_password_hash is required to limit API access for incoming connections to the Listener. The username is
-    -- "source-${id}", allowing an early verification before having to parse the POSTed event.
-    --
-    -- This behavior might change in the future to become "type"-dependable.
-    listener_password_hash text NOT NULL,
+    -- The column listener_password_hash is type-dependent.
+    -- If type is not "icinga2", listener_password_hash is required to limit API access for incoming connections
+    -- to the Listener. The username will be "source-${id}", allowing early verification.
+    listener_password_hash text,
+
+    -- Following columns are for the "icinga2" type.
+    -- At least icinga2_base_url, icinga2_auth_user, and icinga2_auth_pass are required - see CHECK below.
+    icinga2_base_url text,
+    icinga2_auth_user text,
+    icinga2_auth_pass text,
+    -- icinga2_ca_pem specifies a custom CA to be used in the PEM format, if not NULL.
+    icinga2_ca_pem text,
+    -- icinga2_common_name requires Icinga 2's certificate to hold this Common Name if not NULL. This allows using a
+    -- differing Common Name - maybe an Icinga 2 Endpoint object name - from the FQDN within icinga2_base_url.
+    icinga2_common_name text,
+    icinga2_insecure_tls boolenum NOT NULL DEFAULT 'n',
 
     -- The hash is a PHP password_hash with PASSWORD_DEFAULT algorithm, defaulting to bcrypt. This check roughly ensures
     -- that listener_password_hash can only be populated with bcrypt hashes.
     -- https://icinga.com/docs/icinga-web/latest/doc/20-Advanced-Topics/#manual-user-creation-for-database-authentication-backend
-    CHECK (listener_password_hash LIKE '$2y$%'),
+    CHECK (listener_password_hash IS NULL OR listener_password_hash LIKE '$2y$%'),
```

Comment:
This was introduced in #132. Please refer to this PR and the comment above this `CHECK`.

As there is no function matching exactly [PHP's generic `password_hash` function](https://www.php.net/manual/en/function.password-hash.php) in Go - especially regarding future PHP changes -, we are verifying against the current default hashing algorithm, bcrypt, in icinga-notifications.

If a new default will be introduced, e.g., Argon2id, we can make the code more dynamic to support different PHP version defaults. But for the moment, this seems useless as icingaweb2 uses `PASSWORD_DEFAULT` and, afaik, no upcoming change to the default was announced.

To make sure that no other hashing algorithm is used, this additional SQL `CHECK` was added. This results in an early failure which is easy to catch. An additional check was also introduced to the Go code just to be sure.

https://github.com/Icinga/icinga-notifications/blob/4e5716e308feb3e954456f7db888c1804ab9c0c4/internal/config/runtime.go#L170-L179

Reply:
Okay. Though, is it intended to necessitate a schema change in such a case?

You handle the error, which can't occur as no such invalid value can be stored. Which makes the error handling redundant. (For a postgres backend at least)

Updating the daemon is much less hassle and I guess you need to support the *old* algorithm anyway, just like Web does (that's why the algorithm is denoted in the hash to begin with), so the check must likely go away anyway.

Reply:
As I perhaps did not phrase this clearly enough: In our Go code, the hash is verified and we simply cannot know what future PHP `password_hash` defaults would be. Not only PHP code uses this value.
That's why we let it fail hard if the default has changed and we failed to patch in time.

Reply:
*hard*, as in no source can be added or it's password changed. Okay :sweat_smile:

Reply:
Notifications Web should just use `PASSWORD_BCRYPT` instead of relying on the default when creating a password hash for this column. Anything else would not be understood by the daemon anyways.

Reply:
According to the [docs](https://www.php.net/manual/en/function.password-hash.php), the default was last changed for PHP 5.5, [release in 2013](https://www.php.net/ChangeLog-5.php#5.5.0) (w/o a mention in the change log, of course).

I mean, of course we could lift the SQL `CHECK` but this would result in a Go error, which might be harder to see for an end user. Dunno, what the best behavior should be.

However, this "only" affects the `listener_password_hash` and not the upcoming Icinga 2 API communication.

Reply:
Sorry @nilmerg, I missed your first reply as I almost instantly wrote my second answer and somehow GitHub does not shown yours (or I was blind).

> You handle the error, which can't occur as no such invalid value can be stored. Which makes the error handling redundant. (For a postgres backend at least)

An invalid value can be stored if a future PHP version changes the default, resulting in a hash created by another hashing algorithm being stored in the database.

Thus, we should maybe go with @julianbrost's suggestion and don't rely on a default, but settle for an algorithm.

As I wrote in another answer, the idea of this double error check - I wouldn't call it redundant, it has such a bad connotation - was to let it fail early so that novice users don't have to consult icinga-notification's logs.
Of course, this brings the burden to the web team regarding user feedback.. mea culpa.

> Updating the daemon is much less hassle and I guess you need to support the _old_ algorithm anyway, just like Web does (that's why the algorithm is denoted in the hash to begin with), so the check must likely go away anyway.

Why shouldn't go a daemon update hand-in-hand with a schema update?

Reply:
Of course, with appropriate tooling a schema update isn't much of a deal.

I've went with using `PASSWORD_BCRYPT` now. But only because the actual advantage of `PASSWORD_DEFAULT`, to automatically migrate to a stronger algorithm, is only effectively useful in PHP's own ecosystem. Once dealing with anything outside of that, not relying on this *dynamic upgrade* is the better choice.